[Bug] CompilationErrors displayed for subsequent tasks after the failure of one task

**Feature Request or Bug**
Bug

**Describe the feature request or bug**
I created a Task which had some bugs in it - due to which C# compilation failed on the Grunt when I ran it. Now, when I attempt to execute any other task including a valid shellCmd on a grunt, I am stuck with "CompilerException: CompilationErrors: " errors and the previous C# compilation errors are shown. Is there a way to reset the Grunt so it does not attempt to compile the old task and just executes the specified command?

**To Reproduce**
Steps to reproduce the behavior:
1. Go to Tasks
2. Click on Create
3. Create the task below which has C# bugs in it with a single optional param called 'serverIP': 

```
using SharpUp;
using SharpSploit.Execution;

public static class Task
{
    public static string Execute(string serverIp = "")
    {
        string output = "";
        string command = "";
        try
        {   
            output += "\n"
            output += "\n"
            output += "==== Get hostname via 'hostname' command ===="
            output += "\n"
            output += "\n"
            command = "hostname";
            output += Shell.ShellCmdExecute(command);

            output += "\n"
            output += "\n"
            output += "==== Get current user and permissions via 'whoami' command===="
            output += "\n"
            output += "\n"
            command = "whoami /all";
            output += Shell.ShellCmdExecute(command);

            output += "\n"
            output += "\n"
            output += "=== SharpUp: Running Privilege Escalation Checks similar to PowerUp from Powersploit ==="
            output += "\n"
            output += "\n"
            TextWriter realStdOut = Console.Out;
            TextWriter realStdErr = Console.Error;
            TextWriter stdOutWriter = new StringWriter();
            TextWriter stdErrWriter = new StringWriter();
            Console.SetOut(stdOutWriter);
            Console.SetError(stdErrWriter);
            bool auditMode = args.Contains("audit", StringComparer.CurrentCultureIgnoreCase);
            Program.PrivescChecks(auditMode);
            Console.Out.Flush();
            Console.Error.Flush();
            Console.SetOut(realStdOut);
            Console.SetError(realStdErr);

            output += stdOutWriter.ToString();
            output += stdErrWriter.ToString();
        }
        catch (Exception e) { output += e.GetType().FullName + ": " + e.Message + Environment.NewLine + e.StackTrace; }
        return output;
    }
}
```
4. Now go to interact for a Grunt and run the task above a couple of times.
5. Subsequent execution of any other task including 'WhoAmI' show the compilation errors  for the task above.

**Expected behavior**
The task should abort if it has bugs in it and other tasks should not be affected.

**Screenshots**
Attached below.

**Covenant Server Information:**
 - OS: Kali Linux Rolling
 - Docker or Native: Docker

**Browser Information:**
 - Browser: Firefox 
 - Version: 60.6.3esr(64bit)

**Target Information (System that implant is running on):**
 - OS: Windows 7 SP1
 - Version [e.g. 22]: 6.1.7600

Attached issue log below:
```
[11/26/2019 15:40:27 UTC] MyPrivilegeEscalationChecks aborted

(Covenant) > MyPrivilegeEscalationChecks /serverip:""

CompilerException: CompilationErrors: 
(26,30): error CS0103: The name 'args' does not exist in the current context
(684,34): warning CS0168: The variable 'ex' is declared but never used
(102,42): warning CS0649: Field 'Program.TOKEN_PRIVILEGES.Privileges' is never assigned to, and will always have its default value null
(100,27): warning CS0649: Field 'Program.TOKEN_PRIVILEGES.PrivilegeCount' is never assigned to, and will always have its default value 0

[11/26/2019 15:44:02 UTC] MyPrivilegeEscalationChecks aborted

(Covenant) > MyPrivilegeEscalationChecks /serverip:""

CompilerException: CompilationErrors: 
(11,27): error CS1002: ; expected
(12,27): error CS1002: ; expected
(13,70): error CS1002: ; expected
(14,27): error CS1002: ; expected
(15,27): error CS1002: ; expected
(16,33): error CS1002: ; expected
(19,27): error CS1002: ; expected
(20,27): error CS1002: ; expected
(21,87): error CS1002: ; expected
(22,27): error CS1002: ; expected
(23,27): error CS1002: ; expected
(24,36): error CS1002: ; expected
(27,27): error CS1002: ; expected
(28,27): error CS1002: ; expected
(29,113): error CS1002: ; expected
(30,27): error CS1002: ; expected
(31,27): error CS1002: ; expected
(2,7): error CS0246: The type or namespace name 'SharpSploit' could not be found (are you missing a using directive or an assembly reference?)
(16,13): error CS0103: The name 'command' does not exist in the current context
(17,23): error CS0103: The name 'Shell' does not exist in the current context
(17,45): error CS0103: The name 'command' does not exist in the current context
(24,13): error CS0103: The name 'command' does not exist in the current context
(25,23): error CS0103: The name 'Shell' does not exist in the current context
(25,45): error CS0103: The name 'command' does not exist in the current context
(32,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(32,37): error CS0103: The name 'Console' does not exist in the current context
(33,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(33,37): error CS0103: The name 'Console' does not exist in the current context
(34,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(34,43): error CS0246: The type or namespace name 'StringWriter' could not be found (are you missing a using directive or an assembly reference?)
(35,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(35,43): error CS0246: The type or namespace name 'StringWriter' could not be found (are you missing a using directive or an assembly reference?)
(36,13): error CS0103: The name 'Console' does not exist in the current context
(37,13): error CS0103: The name 'Console' does not exist in the current context
(39,53): error CS0103: The name 'StringComparer' does not exist in the current context
(39,35): error CS1061: 'string[]' does not contain a definition for 'Contains' and no extension method 'Contains' accepting a first argument of type 'string[]' could be found (are you missing a using directive or an assembly reference?)
(41,13): error CS0103: The name 'Console' does not exist in the current context
(42,13): error CS0103: The name 'Console' does not exist in the current context
(43,13): error CS0103: The name 'Console' does not exist in the current context
(44,13): error CS0103: The name 'Console' does not exist in the current context
(49,16): error CS0246: The type or namespace name 'Exception' could not be found (are you missing a using directive or an assembly reference?)
(49,83): error CS0103: The name 'Environment' does not exist in the current context
(684,34): warning CS0168: The variable 'ex' is declared but never used
(100,27): warning CS0649: Field 'Program.TOKEN_PRIVILEGES.PrivilegeCount' is never assigned to, and will always have its default value 0
(102,42): warning CS0649: Field 'Program.TOKEN_PRIVILEGES.Privileges' is never assigned to, and will always have its default value null

[11/26/2019 15:45:02 UTC] MyPrivilegeEscalationChecks aborted

(Covenant) > MyPrivilegeEscalationChecks /serverip:""

CompilerException: CompilationErrors: 
(2,7): error CS0246: The type or namespace name 'SharpSploit' could not be found (are you missing a using directive or an assembly reference?)
(16,13): error CS0103: The name 'command' does not exist in the current context
(17,23): error CS0103: The name 'Shell' does not exist in the current context
(17,45): error CS0103: The name 'command' does not exist in the current context
(24,13): error CS0103: The name 'command' does not exist in the current context
(25,23): error CS0103: The name 'Shell' does not exist in the current context
(25,45): error CS0103: The name 'command' does not exist in the current context
(32,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(32,37): error CS0103: The name 'Console' does not exist in the current context
(33,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(33,37): error CS0103: The name 'Console' does not exist in the current context
(34,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(34,43): error CS0246: The type or namespace name 'StringWriter' could not be found (are you missing a using directive or an assembly reference?)
(35,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(35,43): error CS0246: The type or namespace name 'StringWriter' could not be found (are you missing a using directive or an assembly reference?)
(36,13): error CS0103: The name 'Console' does not exist in the current context
(37,13): error CS0103: The name 'Console' does not exist in the current context
(39,53): error CS0103: The name 'StringComparer' does not exist in the current context
(39,35): error CS1061: 'string[]' does not contain a definition for 'Contains' and no extension method 'Contains' accepting a first argument of type 'string[]' could be found (are you missing a using directive or an assembly reference?)
(41,13): error CS0103: The name 'Console' does not exist in the current context
(42,13): error CS0103: The name 'Console' does not exist in the current context
(43,13): error CS0103: The name 'Console' does not exist in the current context
(44,13): error CS0103: The name 'Console' does not exist in the current context
(49,16): error CS0246: The type or namespace name 'Exception' could not be found (are you missing a using directive or an assembly reference?)
(49,83): error CS0103: The name 'Environment' does not exist in the current context
(684,34): warning CS0168: The variable 'ex' is declared but never used
(102,42): warning CS0649: Field 'Program.TOKEN_PRIVILEGES.Privileges' is never assigned to, and will always have its default value null
(100,27): warning CS0649: Field 'Program.TOKEN_PRIVILEGES.PrivilegeCount' is never assigned to, and will always have its default value 0

[11/26/2019 15:45:28 UTC] Command submitted

(Covenant) > MyPrivilegeEscalationChecks /serverip:""

[11/26/2019 15:45:54 UTC] Command submitted

(Covenant) > MyPrivilegeEscalationChecks /serverip:""

[11/26/2019 15:46:46 UTC] Command submitted

(Covenant) > MyPrivilegeEscalationChecks

[!] Syntax tree already present
Parameter name: trees[45]
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.AddSyntaxTrees(IEnumerable`1 trees)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.Create(String assemblyName, CSharpCompilationOptions options, IEnumerable`1 syntaxTrees, IEnumerable`1 references, CSharpCompilation previousSubmission, Type returnType, Type hostObjectType, Boolean isSubmission)
   at Covenant.Core.Compiler.CompileCSharp(CompilationRequest request) in /app/Core/Compiler.cs:line 112
   at Covenant.Models.Grunts.GruntTask.Compile() in /app/Models/Grunts/GruntTask.cs:line 171
   at Covenant.Models.CovenantContext.CreateGruntTasking(GruntTasking tasking, IHubContext`1 _grunthub) in /app/Models/CovenantContext.cs:line 1951
   at Covenant.Core.Interaction.StartTask(Grunt grunt, GruntTask task, GruntCommand command) in /app/Core/Interact.cs:line 373
   at Covenant.Core.Interaction.Input(CovenantUser user, Grunt grunt, String UserInput) in /app/Core/Interact.cs:line 278

[11/26/2019 15:47:06 UTC] ShellCmd aborted

(Covenant) > ShellCmd

[11/26/2019 15:49:47 UTC] Command submitted

(Covenant) > MyPrivilegeEscalationChecks

[!] Syntax tree already present
Parameter name: trees[45]
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.AddSyntaxTrees(IEnumerable`1 trees)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.Create(String assemblyName, CSharpCompilationOptions options, IEnumerable`1 syntaxTrees, IEnumerable`1 references, CSharpCompilation previousSubmission, Type returnType, Type hostObjectType, Boolean isSubmission)
   at Covenant.Core.Compiler.CompileCSharp(CompilationRequest request) in /app/Core/Compiler.cs:line 112
   at Covenant.Models.Grunts.GruntTask.Compile() in /app/Models/Grunts/GruntTask.cs:line 171
   at Covenant.Models.CovenantContext.CreateGruntTasking(GruntTasking tasking, IHubContext`1 _grunthub) in /app/Models/CovenantContext.cs:line 1951
   at Covenant.Core.Interaction.StartTask(Grunt grunt, GruntTask task, GruntCommand command) in /app/Core/Interact.cs:line 373
   at Covenant.Core.Interaction.Input(CovenantUser user, Grunt grunt, String UserInput) in /app/Core/Interact.cs:line 278

[11/26/2019 15:49:58 UTC] ShellCmd aborted

(Covenant) > ShellCmd

[11/26/2019 15:50:12 UTC] ShellCmd aborted

(Covenant) > ShellCmd /shellcommand:"whoami"

CompilerException: CompilationErrors: 
(10,27): error CS1002: ; expected
(11,27): error CS1002: ; expected
(12,70): error CS1002: ; expected
(13,27): error CS1002: ; expected
(14,27): error CS1002: ; expected
(15,33): error CS1002: ; expected
(18,27): error CS1002: ; expected
(19,27): error CS1002: ; expected
(20,87): error CS1002: ; expected
(21,27): error CS1002: ; expected
(22,27): error CS1002: ; expected
(23,36): error CS1002: ; expected
(26,27): error CS1002: ; expected
(27,27): error CS1002: ; expected
(28,113): error CS1002: ; expected
(29,27): error CS1002: ; expected
(30,27): error CS1002: ; expected
(15,13): error CS0103: The name 'command' does not exist in the current context
(16,45): error CS0103: The name 'command' does not exist in the current context
(23,13): error CS0103: The name 'command' does not exist in the current context
(24,45): error CS0103: The name 'command' does not exist in the current context
(31,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(31,37): error CS0103: The name 'Console' does not exist in the current context
(32,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(32,37): error CS0103: The name 'Console' does not exist in the current context
(33,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(33,43): error CS0246: The type or namespace name 'StringWriter' could not be found (are you missing a using directive or an assembly reference?)
(34,13): error CS0246: The type or namespace name 'TextWriter' could not be found (are you missing a using directive or an assembly reference?)
(34,43): error CS0246: The type or namespace name 'StringWriter' could not be found (are you missing a using directive or an assembly reference?)
(35,13): error CS0103: The name 'Console' does not exist in the current context
(36,13): error CS0103: The name 'Console' does not exist in the current context
(37,30): error CS0103: The name 'args' does not exist in the current context
(37,53): error CS0103: The name 'StringComparer' does not exist in the current context
(38,13): error CS0103: The name 'Program' does not exist in the current context
(39,13): error CS0103: The name 'Console' does not exist in the current context
(40,13): error CS0103: The name 'Console' does not exist in the current context
(41,13): error CS0103: The name 'Console' does not exist in the current context
(42,13): error CS0103: The name 'Console' does not exist in the current context
(47,16): error CS0246: The type or namespace name 'Exception' could not be found (are you missing a using directive or an assembly reference?)
(47,83): error CS0103: The name 'Environment' does not exist in the current context
```

![Screenshot from 2019-11-27 03-28-42](https://user-images.githubusercontent.com/2260634/69655055-e97d5f00-10c9-11ea-9bbf-abd38416455a.png)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 - Fixed missing http profiles on Ubuntu w/ workaround due to corefx issue
 - Fix GET /api/gruntcommand/{id} endpoint missing output
+- Fix [#122][issue-122] multiple connection addresses issue
 
 ## [v0.4] - 2019-10-30
 ### Added
@@ -186,3 +187,5 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 [v0.3.1]: https://github.com/cobbr/Covenant/compare/v0.3...v0.3.1
 [v0.3.2]: https://github.com/cobbr/Covenant/compare/v0.3.1...v0.3.2
 [v0.4]: https://github.com/cobbr/Covenant/compare/v0.3.2...v0.4
+
+[issue-122]: https://github.com/cobbr/Covenant/issues/122

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,9 +11,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 - Improved ComputerName parsing and output for Domain tasks
+- Made changes necessary for C3 integration, allowing outbound SMB grunts
 
 ### Fixed
 - Fixed missing http profiles on Ubuntu w/ workaround due to corefx issue
+- Fix GET /api/gruntcommand/{id} endpoint missing output
 
 ## [v0.4] - 2019-10-30
 ### Added

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Fixed missing http profiles on Ubuntu w/ workaround due to corefx issue
 - Fix GET /api/gruntcommand/{id} endpoint missing output
 - Fix [#122][issue-122] multiple connection addresses issue
+- Fix [#137][issue-137] grunt last checked-in field sorting issue on table
 
 ## [v0.4] - 2019-10-30
 ### Added
@@ -189,3 +190,4 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 [v0.4]: https://github.com/cobbr/Covenant/compare/v0.3.2...v0.4
 
 [issue-122]: https://github.com/cobbr/Covenant/issues/122
+[issue-137]: https://github.com/cobbr/Covenant/issues/137

--- a/Covenant/Views/Grunt/_GruntTable.cshtml
+++ b/Covenant/Views/Grunt/_GruntTable.cshtml
@@ -33,7 +33,7 @@
                             <td>@Html.DisplayFor(modelItem => item.Hostname)</td>
                             <td>@Html.DisplayFor(modelItem => item.UserName)</td>
                             <td>@Html.DisplayFor(modelItem => item.Status)</td>
-                            <td>@Html.DisplayFor(modelItem => item.LastCheckIn)</td>
+                            <td><span style="display: none">@item.LastCheckIn.Ticks</span>@Html.DisplayFor(modelItem => item.LastCheckIn)</td>
                             <td>@Html.DisplayFor(modelItem => item.Integrity)</td>
                             <td>@Html.DisplayFor(modelItem => item.OperatingSystem)</td>
                             <td>@Html.DisplayFor(modelItem => item.Process)</td>
@@ -51,7 +51,7 @@
                             <td>@Html.DisplayFor(modelItem => item.Hostname)</td>
                             <td>@Html.DisplayFor(modelItem => item.UserName)</td>
                             <td>@Html.DisplayFor(modelItem => item.Status)</td>
-                            <td>@Html.DisplayFor(modelItem => item.LastCheckIn)</td>
+                            <td><span style="display: none">@item.LastCheckIn.Ticks</span>@Html.DisplayFor(modelItem => item.LastCheckIn)</td>
                             <td>@Html.DisplayFor(modelItem => item.Integrity)</td>
                             <td>@Html.DisplayFor(modelItem => item.OperatingSystem)</td>
                             <td>@Html.DisplayFor(modelItem => item.Process)</td>

--- a/Covenant/Views/Listener/Create.cshtml
+++ b/Covenant/Views/Listener/Create.cshtml
@@ -43,7 +43,7 @@
     <script type="text/javascript">
         $(document).ready(function () {
             var addAddress = function () {
-                var num = $(".address-row").length;
+                var num = $(".address-row").length - 1;
                 var newitem = '<div class="d-flex flex-row address-row">' +
                               '    <div class="form-group col-md-4 pl-0 mt-1 mb-1">' +
                               '        <input id="ConnectAddresses_' + num + '_" name="ConnectAddresses[' + num + ']" class="form-control address">' +
@@ -130,7 +130,7 @@
             $('.remove-connectaddress-button').click(removeAddress);
 
             $("#http-tab").tab('show');
-            
+
             swapOptions();
             setUrl();
         });

